From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fabian=20Gr=C3=BCnbichler?= <f.gruenbichler@proxmox.com>
Date: Fri, 19 Dec 2025 08:46:26 +0100
Subject: [PATCH] Revert "tcp: always seek for minimal rtt in
 tcp_rcv_rtt_update()"

This reverts commit b879dcb1aeeca278eacaac0b1e2425b1c7599f9f.
---
 net/ipv4/tcp_input.c | 22 ++++++++++++++--------
 1 file changed, 14 insertions(+), 8 deletions(-)

diff --git a/net/ipv4/tcp_input.c b/net/ipv4/tcp_input.c
index e2895a9e621d55db10b5685972f6c8eacdb7aee9..5d07ccca645e0b831b2d1e3aa9e5365e59b7c047 100644
--- a/net/ipv4/tcp_input.c
+++ b/net/ipv4/tcp_input.c
@@ -664,12 +664,10 @@ EXPORT_IPV6_MOD(tcp_initialize_rcv_mss);
  */
 static void tcp_rcv_rtt_update(struct tcp_sock *tp, u32 sample, int win_dep)
 {
-	u32 new_sample, old_sample = tp->rcv_rtt_est.rtt_us;
-	long m = sample << 3;
+	u32 new_sample = tp->rcv_rtt_est.rtt_us;
+	long m = sample;
 
-	if (old_sample == 0 || m < old_sample) {
-		new_sample = m;
-	} else {
+	if (new_sample != 0) {
 		/* If we sample in larger samples in the non-timestamp
 		 * case, we could grossly overestimate the RTT especially
 		 * with chatty applications or bulk transfer apps which
@@ -680,9 +678,17 @@ static void tcp_rcv_rtt_update(struct tcp_sock *tp, u32 sample, int win_dep)
 		 * else with timestamps disabled convergence takes too
 		 * long.
 		 */
-		if (win_dep)
-			return;
-		new_sample = old_sample - (old_sample >> 3) + sample;
+		if (!win_dep) {
+			m -= (new_sample >> 3);
+			new_sample += m;
+		} else {
+			m <<= 3;
+			if (m < new_sample)
+				new_sample = m;
+		}
+	} else {
+		/* No previous measure. */
+		new_sample = m << 3;
 	}
 
 	tp->rcv_rtt_est.rtt_us = new_sample;
